File: server/stripe/service.ts
import type { BillingApi, ProviderFactResult } from '../billing/public';
import {
  STRIPE_MAX_WEBHOOK_BYTES,
  STRIPE_PROVIDER,
  decodeStripeCheckoutResponse,
  decodeStripeEventPlan,
  decodeStripeReconciliationSnapshot,
  planStripeCheckout,
  type StripeSnapshotPlan,
} from './core';
import type { StripeTransportPort, StripeWebhookVerifierPort } from './ports';
import type {
  HostedCheckoutResult,
  StripeBillingAdapter,
  StripeBillingConfiguration,
  StripeWebhookResult,
} from './public';

export function createStripeBillingAdapter(dependencies: {
  readonly configuration: StripeBillingConfiguration;
  readonly billing: BillingApi;
  readonly transport: StripeTransportPort;
  readonly webhookVerifier: StripeWebhookVerifierPort;
}): StripeBillingAdapter {
  return {
    async beginHostedCheckout(context, command): Promise<HostedCheckoutResult> {
      let began: Awaited<ReturnType<BillingApi['beginCheckout']>>;
      try {
        began = await dependencies.billing.beginCheckout(context, {
          subscriptionId: command.subscriptionId,
          checkoutIntentId: command.checkoutIntentId,
          provider: STRIPE_PROVIDER,
          createdAt: command.createdAt,
        });
      } catch {
        return { kind: 'rejected', reason: 'billing-rejected' };
      }
      if (began.kind === 'rejected') {
        return {
          kind: 'rejected',
          reason:
            began.reason === 'invalid-transition'
              ? 'invalid-input'
              : 'billing-rejected',
        };
      }

      let rawResponse: unknown;
      try {
        rawResponse = await dependencies.transport.createCheckoutSession(
          planStripeCheckout(dependencies.configuration, command),
        );
      } catch {
        return { kind: 'rejected', reason: 'provider-unavailable' };
      }
      const response = decodeStripeCheckoutResponse(rawResponse, {
        mode: dependencies.configuration.mode,
        subscriptionId: command.subscriptionId,
        checkoutIntentId: command.checkoutIntentId,
      });
      if (response === undefined) {
        return { kind: 'rejected', reason: 'malformed-provider-response' };
      }
      let opened: Awaited<ReturnType<BillingApi['recordCheckoutOpened']>>;
      try {
        opened = await dependencies.billing.recordCheckoutOpened(context, {
          subscriptionId: command.subscriptionId,
          checkoutIntentId: command.checkoutIntentId,
          providerCheckoutReference: response.providerCheckoutReference,
          openedAt: command.createdAt,
        });
      } catch {
        return { kind: 'rejected', reason: 'billing-rejected' };
      }
      if (opened.kind === 'rejected') {
        return {
          kind: 'rejected',
          reason:
            opened.reason === 'identifier-conflict'
              ? 'provider-mapping-mismatch'
              : 'billing-rejected',
        };
      }
      return {
        kind: 'redirect',
        checkoutUrl: response.checkoutUrl,
        providerCheckoutReference: response.providerCheckoutReference,
      };
    },

    async ingestWebhook(request): Promise<StripeWebhookResult> {
      if (
        !(request.rawBody instanceof Uint8Array) ||
        request.rawBody.byteLength === 0 ||
        request.rawBody.byteLength > STRIPE_MAX_WEBHOOK_BYTES
      ) {
        return { kind: 'rejected', reason: 'malformed-event' };
      }
      let verification;
      try {
        verification = await dependencies.webhookVerifier.verify(request);
      } catch {
        return { kind: 'rejected', reason: 'invalid-signature' };
      }
      if (verification.kind === 'rejected') {
        return { kind: 'rejected', reason: 'invalid-signature' };
      }
      const payload = parseJsonUtf8(verification.rawBody);
      if (payload.kind === 'rejected') return payload;
      const plan = decodeStripeEventPlan(payload.value, {
        mode: dependencies.configuration.mode,
        apiVersion: dependencies.configuration.apiVersion,
        receivedAt: request.receivedAt,
      });
      switch (plan.kind) {
        case 'unsupported':
          return { kind: 'ignored', reason: 'unsupported-event' };
        case 'rejected':
          return plan;
        case 'fact':
          return ingestFact(dependencies.billing, plan.fact);
        case 'snapshot':
          return retrieveAndReconcile(dependencies, plan.plan);
      }
    },

    async reconcileSubscription(command): Promise<StripeWebhookResult> {
      return retrieveAndReconcile(dependencies, {
        snapshotId: command.snapshotId,
        subscriptionId: command.subscriptionId,
        providerSubscriptionReference: command.providerSubscriptionReference,
        observedAt: command.observedAt,
        recordedAt: command.recordedAt,
      });
    },
  };
}

async function retrieveAndReconcile(
  dependencies: {
    readonly configuration: StripeBillingConfiguration;
    readonly billing: BillingApi;
    readonly transport: StripeTransportPort;
  },
  plan: StripeSnapshotPlan,
): Promise<StripeWebhookResult> {
  let input: unknown;
  try {
    input = await dependencies.transport.retrieveSubscriptionSnapshot({
      apiVersion: dependencies.configuration.apiVersion,
      providerSubscriptionReference: plan.providerSubscriptionReference,
    });
  } catch {
    return { kind: 'rejected', reason: 'provider-unavailable' };
  }
  const snapshot = decodeStripeReconciliationSnapshot(input, plan);
  if (snapshot === undefined) {
    return { kind: 'rejected', reason: 'malformed-event' };
  }
  try {
    return providerResult(
      await dependencies.billing.reconcileVerifiedSnapshot(snapshot),
    );
  } catch {
    return { kind: 'rejected', reason: 'billing-rejected' };
  }
}

async function ingestFact(
  billing: BillingApi,
  fact: Parameters<BillingApi['ingestVerifiedProviderFact']>[0],
): Promise<StripeWebhookResult> {
  try {
    return providerResult(await billing.ingestVerifiedProviderFact(fact));
  } catch {
    return { kind: 'rejected', reason: 'billing-rejected' };
  }
}

function providerResult(result: ProviderFactResult): StripeWebhookResult {
  return result.kind === 'rejected'
    ? { kind: 'rejected', reason: 'billing-rejected' }
    : { kind: 'accepted', outcome: result.kind };
}

function parseJsonUtf8(
  rawBody: Uint8Array,
):
  | { readonly kind: 'parsed'; readonly value: unknown }
  | { readonly kind: 'rejected'; readonly reason: 'malformed-event' } {
  try {
    const text = new TextDecoder('utf-8', {
      fatal: true,
      ignoreBOM: false,
    }).decode(rawBody);
    const value: unknown = JSON.parse(text);
    return { kind: 'parsed', value };
  } catch {
    return { kind: 'rejected', reason: 'malformed-event' };
  }
}

File: backend/internal/stripebilling/service.go
package stripebilling

import (
	"context"
	"errors"

	"github.com/fukamu/notes/backend/internal/billing"
	"github.com/fukamu/notes/backend/internal/identity"
)

var ErrInvalidServiceConfiguration = errors.New("invalid Stripe billing service configuration")

type Service struct {
	configuration Configuration
	billing       BillingPort
	provider      ProviderPort
	verifier      WebhookVerifierPort
}

func NewService(configuration Configuration, billingPort BillingPort, provider ProviderPort, verifier WebhookVerifierPort) (*Service, error) {
	if !configuration.Valid() || billingPort == nil || provider == nil || verifier == nil {
		return nil, ErrInvalidServiceConfiguration
	}
	return &Service{configuration: configuration, billing: billingPort, provider: provider, verifier: verifier}, nil
}

func (service *Service) BeginHostedCheckout(
	ctx context.Context,
	vaultContext identity.VaultContext,
	command HostedCheckoutCommand,
) HostedCheckoutResult {
	planned, err := PlanCheckout(service.configuration, command)
	if err != nil {
		return rejectedCheckout(ReasonInvalidInput)
	}
	began, err := service.billing.BeginCheckout(ctx, vaultContext, billing.BeginCheckoutCommand{
		SubscriptionID: command.SubscriptionID, CheckoutID: command.CheckoutIntentID,
		Provider: Provider, CreatedAt: command.CreatedAt,
	})
	if err != nil {
		return rejectedCheckout(ReasonBillingRejected)
	}
	if began.Kind == billing.ResultRejected {
		if began.Reason == billing.ResultInvalidTransition {
			return rejectedCheckout(ReasonInvalidInput)
		}
		return rejectedCheckout(ReasonBillingRejected)
	}
	providerResponse, err := service.provider.CreateCheckoutSession(ctx, planned)
	if err != nil {
		return rejectedCheckout(ReasonProviderUnavailable)
	}
	response, reason := DecodeCheckoutResponse(providerResponse, command, service.configuration.Mode)
	if reason != "" {
		return rejectedCheckout(reason)
	}
	opened, err := service.billing.RecordCheckoutOpened(ctx, vaultContext, billing.RecordCheckoutOpenedCommand{
		SubscriptionID: command.SubscriptionID, CheckoutIntentID: command.CheckoutIntentID,
		ProviderCheckoutReference: response.ProviderCheckoutReference, OpenedAt: command.CreatedAt,
	})
	if err != nil {
		return rejectedCheckout(ReasonBillingRejected)
	}
	if opened.Kind == billing.ResultRejected {
		if opened.Reason == billing.ResultIdentifierConflict {
			return rejectedCheckout(ReasonProviderMappingMismatch)
		}
		return rejectedCheckout(ReasonBillingRejected)
	}
	return HostedCheckoutResult{Kind: HostedCheckoutRedirect, CheckoutURL: response.CheckoutURL, ProviderCheckoutReference: response.ProviderCheckoutReference}
}

func (service *Service) IngestWebhook(ctx context.Context, request WebhookRequest) WebhookResult {
	if len(request.RawBody) == 0 || len(request.RawBody) > MaxWebhookBytes || !validMillis(request.ReceivedAt) {
		return rejectedWebhook(ReasonMalformedEvent)
	}
	verified := service.verifier.Verify(request)
	if !verified.Verified {
		return rejectedWebhook(ReasonInvalidSignature)
	}
	plan := DecodeEventPlan(verified.RawBody, service.configuration.Mode, service.configuration.APIVersion, request.ReceivedAt)
	switch plan.Kind {
	case EventUnsupported:
		return WebhookResult{Kind: WebhookIgnored}
	case EventRejected:
		return rejectedWebhook(plan.Reason)
	case EventFact:
		if plan.Fact == nil {
			return rejectedWebhook(ReasonMalformedEvent)
		}
		result, err := service.billing.IngestVerifiedProviderFact(ctx, *plan.Fact)
		return billingResult(result, err)
	case EventSnapshot:
		if plan.Snapshot == nil {
			return rejectedWebhook(ReasonMalformedEvent)
		}
		return service.retrieveAndReconcile(ctx, *plan.Snapshot)
	default:
		return rejectedWebhook(ReasonMalformedEvent)
	}
}

func (service *Service) ReconcileSubscription(ctx context.Context, command ReconciliationCommand) WebhookResult {
	return service.retrieveAndReconcile(ctx, SnapshotPlan{
		SnapshotID: command.SnapshotID, SubscriptionID: command.SubscriptionID,
		ProviderSubscriptionReference: command.ProviderSubscriptionReference,
		ObservedAt:                    command.ObservedAt, RecordedAt: command.RecordedAt,
	})
}

func (service *Service) retrieveAndReconcile(ctx context.Context, plan SnapshotPlan) WebhookResult {
	input, err := service.provider.RetrieveSubscriptionSnapshot(ctx, plan.ProviderSubscriptionReference)
	if err != nil {
		return rejectedWebhook(ReasonProviderUnavailable)
	}
	snapshot, ok := DecodeReconciliationSnapshot(input, plan)
	if !ok {
		return rejectedWebhook(ReasonMalformedEvent)
	}
	result, err := service.billing.ReconcileVerifiedSnapshot(ctx, snapshot)
	return billingResult(result, err)
}

func billingResult(result billing.CommandResult, err error) WebhookResult {
	if err != nil || result.Kind == billing.ResultRejected {
		return rejectedWebhook(ReasonBillingRejected)
	}
	return WebhookResult{Kind: WebhookAccepted, Outcome: result.Kind}
}

func rejectedCheckout(reason RejectionReason) HostedCheckoutResult {
	return HostedCheckoutResult{Kind: HostedCheckoutRejected, Reason: reason}
}

func rejectedWebhook(reason RejectionReason) WebhookResult {
	return WebhookResult{Kind: WebhookRejected, Reason: reason}
}
